src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -0,0 +1,207 @@
+            // Now retry
+            error = mAppServer->GetCASESessionManager()->FindOrEstablishSession(peer, &mOnConnectedCallback,
+                                                                                &mOnConnectionFailureCallback);
+        }
+    }
+    return error;
+}
+
+CHIP_ERROR BindingManager::EnqueueUnicastNotification(FabricIndex fabric, NodeId node, EndpointId endpoint, ClusterId cluster,
+                                                      void * context)
+{
+    VerifyOrReturnError(mAppServer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    FabricInfo * fabricInfo = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
+    VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_NOT_FOUND);
+    PeerId peer = fabricInfo->GetPeerIdForNode(node);
+    return mPendingNotificationMap.AddPendingNotification(peer, endpoint, cluster, context);
+}
+
+void BindingManager::HandleDeviceConnected(void * context, OperationalDeviceProxy * device)
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnected(device);
+}
+
+void BindingManager::HandleDeviceConnected(OperationalDeviceProxy * device)
+{
+    mPendingNotificationMap.ForEachActiveObject([&](PendingNotificationEntry * entry) -> Loop {
+        if (entry->GetPeerId() == device->GetPeerId())
+        {
+            SyncPendingNotificationsToPeer(device, entry);
+        }
+
+        return Loop::Continue;
+    });
+}
+
+void BindingManager::SyncPendingNotificationsToPeer(OperationalDeviceProxy * device, PendingNotificationEntry * pendingClusters)
+{
+    for (const ClusterPath & path : *pendingClusters)
+    {
+        ClusterId cluster   = path.cluster;
+        EndpointId endpoint = path.endpoint;
+        for (uint8_t j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
+        {
+            EmberBindingTableEntry entry;
+            if (emberGetBinding(j, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.clusterId == cluster &&
+                entry.local == endpoint && mBoundDeviceChangedHandler)
+            {
+                mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, device, NullOptional, path.context);
+            }
+        }
+    }
+    mPendingNotificationMap.RemoveEntry(pendingClusters);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(void * context, PeerId peerId, CHIP_ERROR error)
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnectionFailure(peerId, error);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(PeerId peerId, CHIP_ERROR error)
+{
+    // Simply release the entry, the connection will be re-established as needed.
+    ChipLogError(AppServer, "Failed to establish connection to node 0x" ChipLogFormatX64, ChipLogValueX64(peerId.GetNodeId()));
+    mAppServer->GetCASESessionManager()->ReleaseSession(peerId);
+}
+
+CHIP_ERROR BindingManager::RemoveUnicastBinding(FabricIndex fabric, NodeId node)
```

Comment:
This should be named LastUnicastBindingRemoved, no?  That's what this notification really means.